gobblin-compaction/src/main/java/gobblin/compaction/verify/DataCompletenessVerifier.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.compaction.verify;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import gobblin.compaction.Dataset;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ExecutorsUtils;
+
+
+/**
+ * A class for verifying data completeness of a {@link Dataset}.
+ *
+ * To verify data completeness, one should extend {@link AbstractRunner} and implement {@link AbstractRunner#call()}
+ * which returns a {@link Results} object. The (fully qualified) name of the class that extends {@link AbstractRunner}
+ * should be associated with property {@link ConfigurationKeys#COMPACTION_DATA_COMPLETENESS_VERIFICATION_CLASS}.
+ *
+ * @author ziliu
+ */
+public class DataCompletenessVerifier implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DataCompletenessVerifier.class);
+
+  private final State props;
+  private final ListeningExecutorService exeSvc;
+
+  /**
+   * The given {@link State} object must specify property
+   * {@link ConfigurationKeys#COMPACTION_DATA_COMPLETENESS_VERIFICATION_CLASS}, and may optionally specify
+   * {@link ConfigurationKeys#COMPACTION_DATA_COMPLETENESS_VERIFICATION_THREAD_POOL_SIZE}.
+   */
+  public DataCompletenessVerifier(State props) {
+    this.props = props;
+    this.exeSvc = getExecutorService();
+  }
+
+  private ListeningExecutorService getExecutorService() {
+    int threadPoolSize = getDataCompletenessVerificationThreadPoolSize();
+    return MoreExecutors.listeningDecorator(Executors.newFixedThreadPool(threadPoolSize));
```

Comment:
Using a fixed thread pools ties up resources (threads) when not needed. Have a look at gobblin.util.executors.ScalingThreadPoolExecutor . We should use a thread pool with sizes (0, threadPoolSize) so that the thread pool can grow and shrink as needed.